File: Binary Search/boj_2805.py
n, m = map(int, input().split())
trees = list(map(int, input().split()))

start, end = 1, max(trees)

while start <= end:
    mid = (start + end) // 2
    height = 0
    for t in trees:
        diff = t - mid
        if diff < 0:
            diff = 0
        height += diff

    if height >= m:
        start = mid + 1
    else:
        end = mid - 1

print(end)

File: Binary Search/boj_2805.cpp
#include <iostream>
#include <algorithm>
using namespace std;

int main() {
	int n, m;
	int trees[1000000];

	cin >> n >> m;
	for (int i = 0; i < n; i++)
		cin >> trees[i];

	int start = 1, end = *max_element(trees, trees + n);
	int mid = 0;

	while (start <= end) {
		long long sum = 0;
		mid = (start + end) / 2;

		for (int i = 0; i < n; i++) {
			if (trees[i] > mid)
				sum += trees[i] - mid;
		}

		if (sum >= m)
			start = mid + 1;
		else
			end = mid - 1;
	}
	
	cout << end;

	return 0;
}
